### PR TITLE
Randomize message processing peer order

### PR DESCRIPTION
**Pull request by sipa**
Right now, the message handling loop iterates the list of nodes always in the same order: the order they were connected in (see the `vNodes` vector). For some parts of the net processing logic, this order matters. Transaction requests are assigned explicitly to peers since #19988, but many other parts of processing work on a "first-served-by-loop-first" basis, such as block downloading. If peers can predict this ordering, it may be exploited to cause delays.

As there isn't anything particularly optimal about the current ordering, just make it unpredictable by randomizing.

Reported by Crypt-iQ.

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

**Comment by theStack** [comment]
Concept ACK

**Comment by prayank23** [comment]
Concept ACK

> Randomize the order in which we process messages from/to our peers. This prevents attacks in which an attacker exploits having multiple consecutive connections in the vNodes list.

I am assuming it improves privacy and security.

**Comment by Crypt-iQ** [comment]
Concept ACK

**Comment by jnewbery** [comment]
ACK 79c02c88b3

We shouldn't build any assumptions into net_processing about the order that we serve peers. Shuffling the order on each loop of the message handler is a good way to make that requirement more explicit.

I think in the future, we want to move away from a strictly equally-weighted algorithm to service our peers in the message handling thread. If a peer is providing us with lots of useful data, we shouldn't have to call ProcessMessages()/SendMessages() for every peer each time we process a message from the useful peer. Likewise, if a peer is causing us to use a lot of resources, we should be able to throttle back on serving it. That would mean a non-stable ordering of servicing peers, which would again mean that net_processing can't assume a stable ordering. Merging this PR should give us confidence that there aren't any undocumented assumptions about ordering in net_processing and that those future changes are safe.

In summary, I think this is a useful change now, and is consistent with the way we should think about the net-net_processing interface.

**Comment by kristapsk** [comment]
Concept ACK

**Comment by Crypt-iQ** [comment]
ACK 79c02c88b347f1408a2db307db2654917f9b0bcc

**Comment by sipa** [comment]
FWIW, shuffling 1000 pointers takes around 10 microseconds here, if anyone was concerned about the CPU cost (in a microbenchmark).

**Review by jonatack — COMMENTED** [review]
> FWIW, shuffling 1000 pointers takes around 10 microseconds here, if anyone was concerned about the CPU cost (in a microbenchmark).

Thanks, I was curious what the impact would be. 

I've been running a node with this patch rebased to master (to also continue testing the torv2 removal) since a few minutes after this was opened, as a sanity check, and all seems nominal.

ACK 79c02c88b347f1408a2db307db2654917f9b0bcc

**Comment by sdaftuar** [comment]
utACK 79c02c88b347f1408a2db307db2654917f9b0bcc

**Review by vasild — APPROVED** [review]
ACK 79c02c88b347f1408a2db307db2654917f9b0bcc

The patch looks harmless and will achieve what it claims. Maybe it is just me, but I don't see the attack vector.

**Comment by luke-jr** [comment]
Is there a risk this could make things worse? Right now, longer-connected peers can't be "disrupted" (in this context) by later-connected peers, but with a random order, disruption opportunities also open up to later peers where they might not have previously been.

**Comment by sipa** [comment]
@luke-jr Interruption isn't changed, we always progress from one peer to the next one in `vNodes` to the next one in the list, at predefined times (when all its messages have been processed, or after certain things like `getdata` for blocks). What this changes is removing the predictability of that order.

**Comment by jamesob** [comment]
Concept ACK

**Review by theStack — APPROVED** [review]
ACK 79c02c88b347f1408a2db307db2654917f9b0bcc

Built and ran the PR branch on mainnet for the last few hours, didn't notice any problems.

**Comment by achow101** [comment]
Code Review ACK 79c02c88b347f1408a2db307db2654917f9b0bcc

**Comment by jamesob** [comment]
crACK https://github.com/bitcoin/bitcoin/pull/22144/commits/79c02c88b347f1408a2db307db2654917f9b0bcc

**Comment by fanquake** [comment]
Edited op to remove @ mention.

**Comment by vasild** [comment]
Should this shuffling also be done in [`CConnman::SocketHandler()`](https://github.com/bitcoin/bitcoin/blob/74013641e035a2f1b12383e63938a5e848506df3/src/net.cpp#L1512)? I will do that in https://github.com/bitcoin/bitcoin/pull/21943 mainly because it is most clean to create the nodes' snapshot shuffled.
